src/ng/parse.js

```diff
@@ -622,15 +622,44 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isPure(node, parentIsPure) {
+  switch (node.type) {
+    // Computed members might invoke a stateful toString()
+    case AST.MemberExpression:
+      if (node.computed) {
+        return false;
+      }
+      break;
+
+    // Unary always convert to primative
+    case AST.UnaryExpression:
+      return true;
```

Comment:
Hm...OK maybe that is not true, because unary operators will always evaluate to the same thing for objects whose `valueOf` returns a non-primitve value (`!` will return `false`, `+/-` will return `NaN`), so this might actually work correctly :grin: